core/drivers/tpm2_mmio.c

```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022, Linaro Limited
+ *
+ */
+
+#include <drivers/tpm2_mmio.h>
+#include <io.h>
+#include <trace.h>
+
+static vaddr_t tpm2_chip_to_base(struct tpm2_chip *chip)
+{
+	struct tpm2_mmio_data *md =
+		container_of(chip, struct tpm2_mmio_data, chip);
+
+	return md->base.va;
+}
+
+static enum tpm2_result tpm2_mmio_rx32(struct tpm2_chip *chip, uint32_t adr,
+				       uint32_t *buf)
+{
+	vaddr_t base = tpm2_chip_to_base(chip);
+
+	*buf = io_read32(base + adr);
+
+	return TPM2_OK;
+}
+
+static enum tpm2_result tpm2_mmio_tx32(struct tpm2_chip *chip, uint32_t adr,
+				       uint32_t val)
+{
+	vaddr_t base = tpm2_chip_to_base(chip);
+
+	io_write32(base + adr, val);
+
+	return TPM2_OK;
+}
+
+static enum tpm2_result tpm2_mmio_rx8(struct tpm2_chip *chip, uint32_t adr,
+				      uint16_t len, uint8_t *buf)
+{
+	vaddr_t base = tpm2_chip_to_base(chip);
+
+	while (len--)
```

Comment:
I'd prefer something less error prone, like:
``` C
for (n = 0; n < len; n++)
         buf[n] = io_read8(base + adr);
```

By the way, you're doing byte reads from the same device mapped address multiple times. What happens if the producer can't keep up? Are the reads just stalling until there's something to read or does something else happens?